internal/mdns/mdns.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package mdns
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/whyrusleeping/mdns"
+)
+
+// Notifee is notified when a new peer is found.
+type Notifee interface {
+	HandlePeerFound(peer.AddrInfo)
+}
+
+// Service implements a mDNS service.
+type Service struct {
+	server     *mdns.Server
+	p2pHost    IDNetworker
+	serviceTag string
+	notifee    Notifee
+	logger     Logger
+	started    bool
+	stop       chan struct{}
+	done       chan struct{}
+}
+
+// NewService creates and returns a new mDNS service.
+func NewService(p2pHost IDNetworker, serviceTag string,
+	logger Logger, notifee Notifee) (service *Service) {
+	if serviceTag == "" {
+		serviceTag = "_ipfs-discovery._udp"
+	}
+
+	return &Service{
+		p2pHost:    p2pHost,
+		serviceTag: serviceTag,
+		notifee:    notifee,
+		logger:     logger,
+	}
+}
+
+// Start starts the mDNS service.
+func (s *Service) Start() (err error) {
+	ips, port := getMDNSIPsAndPort(s.p2pHost)
+
+	hostID := s.p2pHost.ID()
+
+	hostIDPretty := hostID.Pretty()
+	txt := []string{hostIDPretty}
+	mdnsService, err := mdns.NewMDNSService(hostIDPretty, s.serviceTag, "", "", int(port), ips, txt)
+	if err != nil {
+		return fmt.Errorf("creating mDNS service: %w", err)
+	}
+
+	server, err := mdns.NewServer(&mdns.Config{Zone: mdnsService})
+	if err != nil {
+		return fmt.Errorf("creating mDNS server: %w", err)
+	}
+	s.server = server
+
+	s.started = true
+	s.stop = make(chan struct{})
+	s.done = make(chan struct{})
+	ready := make(chan struct{})
+
+	go s.run(ready)
+	// It takes a few milliseconds to launch a goroutine
+	// so we wait for the run goroutine to be ready.
+	<-ready
+
+	return nil
+}
+
+// Stop stops the mDNS service and server.
+func (s *Service) Stop() (err error) {
+	if !s.started {
+		return nil
```

Comment:
we should probably return an error in this case?

Reply:
Actually that's a good question, what should a service implementer do if

- the service is stopped but was already stopped
- the service is started but was already started

Possible solutions:

1. just return nil, since the state transition is already done
2. return an error
3. panic, since it's a programming error on the caller side
4. not care about the state of the service

I think 4. is risky and would lead to possibly working but unexpected behavior, although it's easier for a service implementer. I would rather do 3. than 2., since really it's a programming error on the caller, not the service failing to stop due to an error (otherwise it might got unnoticed). Then 1. works fine (we don't need to crash in the end really), but allows for bad calling code. So in my mind it's 1. or 3., maybe leaning towards 3. Any thoughts?


-

Reply:
Well we can't really ensure that we don't call `Stop` more than once.  So I'd probably just return an error in this case.

Reply:
I don't see the point of an error, it's not something coming from the service failing to start or stop really. It's more of a programming error on the caller. If we want to pay attention to this, we should panic instead so the caller can fix up their code (instead of returning an error which might be just logged and never seen).

I'm also thinking more and more 1. is fine. If we send a network request to change state of something to X, and it's already at state X, I'd say it's fine to return an OK status (=nil error here).

Reply:
I'd advise against panicing this case.  From previous experience on implementing this interface for data pipelines, there could be cases where `Stop()` could be called multiple times.  Take the `SIGINT` to `SIGTERM` example.  If you had something listening on both signals, you could trigger `Service.Stop()` in both cases.

Reply:
Yeah ok, this could make sense. Should we then just return `nil`? We're trying to transition to the same state, I don't think we should really return an error right? :thinking: (haha I know jumping from panic to return nil feels odd)

Reply:
I think returning `nil` is fine.  In what case would you return an error for this service?

Reply:
- When starting fails due to the setup before launching the internal goroutine (setup files, download stuff, check settings, etc.)
- When stopping fails due to the internal goroutine failing to stop or cleanup (remove files, close connections etc.)

I'm very shared between panicing and returning nil, but on the other hand, sending a request to change a state from X to X often succeed, so I guess it's fine to return nil and not panic (leading to possibly worst calling code, but the caller might not care about the service telling them this :smile:)

Reply:
Asked on reddit: https://www.reddit.com/r/golang/comments/101ah0j

Looks like `return nil` is the preferred way so it stays idempotent.